New histogram metric for migration duration

ArthurSens opened the pull request:
**What this PR does / why we need it**:
### New metric  `kubevirt_migration_duration_seconds`
This PR adds a new histogram `kubevirt_migration_duration_seconds` that provides detailed information about time spent on VMI migrations.

Migrations can be aggregated by the following labels:
* `target_node` - Target node of the migration
* `source_node` - Source node of the migration
* `name` - Name of the VMI that is being migrated 
* `namespace` - VMI's namespace
* `status` - Whether the migration was successful or not

`name` and `namespace` should be used together to identify a particular VMI

This new metric can be used in the following use cases:
* Detect if migrations are failing:
```
delta(kubevirt_migration_duration_seconds_count{status=”failed”}[5m]) > 1
```
* Detect average time spent on migrations:
```
kubevirt_migration_duration_seconds_sum / kubevirt_migration_duration_seconds_count
rate(kubevirt_migration_duration_seconds_sum[5m]) / rate(kubevirt_migration_duration_seconds_count[5m]) #average time spent in last 5 minutes
```
* Detect average time spent on migrations, aggregating by any desired label, e.g., `status`:
```
sum by(status) (kubevirt_migration_duration_seconds_sum) / 
sum by(status) (kubevirt_migration_duration_seconds_count)
```
Let's say you have an SLA that says that 90% of migrations should take less than 1h. You can calculate the percentile of migrations duration on the last day and set up an alert  with the following PromQL:
```
histogram_quantile(0.9, 
  rate(kubevirt_migration_duration_seconds_bucket[24h])
) > 3600
```

If your SLA is more flexible and have different agreements for different nodes or VMIs, you can measure percentiles accordingly, e.g., per VMI:
```
histogram_quantile(0.9, 
  sum by(name, namespace)(
    rate(kubevirt_migration_duration_seconds_bucket[24h])
  )
)
```
You may want to be sure that, on the last day, no migrations took more than 1h and to be alerted otherwise:
```
delta(kubevirt_migration_duration_seconds_bucket{"le"="3600"}[24h]) > 1
```
This leads to the second part of this PR

### Metrics configured by the cluster-admin

One may use KubeVirt to deploy really small VMs, leading to really fast VMI migrations. One may also use KubeVirt for really big VMs, leading to migrations that may take a lot of time.

It's really complicated to set up buckets that are useful for every use-case, so this PR also wants to add the possibility to configure histogram buckets through the KubeVirt Custom Resource.

It will add the following API changes:
* Possibility to create buckets that do not follow a linear or exponential progression:
```
apiVersion: kubevirt.io/v1alpha3
kind: KubeVirt
metadata:
  name: kubevirt
spec:
  metricsConfig:
    migrationMetrics: 
      durationHistogram:
        bucketValues: [60, 180, 300, 1800, 3600]
```

### Note:
If left black, will create a custom bucket with the following values:
```
# Values are in seconds and corresponds to: 1m, 5m, 10m, 30m, 1h, 2h, 3h, 4h, 5h, 6h, 7h and 10h
[60, 300, 600, 1800, 3600, 7200, 10800, 14400, 18000, 21600, 25200, 36000]
```


The API changes must be done in a way that it would be easy to add configuration regarding different metrics that may be added in the future. Snapshots and snapshot restores may have different duration when comparing to migration and one may want to set up different buckets. Metrics with different units may also be added in the future, e.g., how much data was transferred in a particular migration.



**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3865 

**Special notes for your reviewer**:
References used to create this PR:
* [How does a prometheus histogram work](https://www.robustperception.io/how-does-a-prometheus-histogram-work)
* [Prometheus client_golang examples](https://github.com/prometheus/client_golang/blob/master/examples/random/main.go)


**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Added new histogram "kubevirt_migration_duration_seconds" to the virt-controller
KubeVirt CR can be used to configure histogram metrics
```

ArthurSens commented on pkg/virt-handler/vm.go:
```diff
@@ -649,6 +649,18 @@ func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstanc
+
+			// Add sample to migration histogram if migration was completed
+			if migrationMetadata.Completed {
+				timeSpent := migrationMetadata.StartTimestamp.Time.Sub(migrationMetadata.EndTimestamp.Time)
+				if migrationMetadata.Failed {
+					// migrationMetrics would come from pkg/monitoring/migrations/prometheus
+					// Not sure where to declare it though
+					migrationMetrics.observeMigrationDuration(float64(timeSpent/time.Second), vmi.Status.MigrationState.SourceNode, vmi.Status.MigrationState.TargetNode, "failed")
+				} else {
+					migrationMetrics.observeMigrationDuration(float64(timeSpent/time.Second), vmi.Status.MigrationState.SourceNode, vmi.Status.MigrationState.TargetNode, "successful")
+				}
+			}
```
This won't compile, of course, because `migrationMetrics` wasn't declared anywhere.

I _think_ that it must be _declared at/added to_ the virt-controller component since I don't want each virt-handler to expose it with different values, but I'm not sure how to do that :grimacing: 

@rmohr @xpivarc @vladikr could you take a look at it when you have some time?

ArthurSens commented:
I've decided to remove the `source` and `target` labels. Now, I don't see too much value being added by them, so I prefer to not add them to reduce storage usage.

Also, if needed in the future, it's easy to re-add them, but it's not so simple to deprecate them once they are released.

ArthurSens commented:
Hi @rmohr, looks like some tests are panicking with the UpdateStatus function, but I'm not really familiar with what it does so I still couldn't debug this

Could you give me a hand and guide me on how to fix this issue?

xpivarc commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -280,32 +285,43 @@ func (c *MigrationController) updateStatus(migration *virtv1.VirtualMachineInsta
 		migrationCopy.Status.Phase = virtv1.MigrationFailed
 		c.recorder.Eventf(migration, k8sv1.EventTypeWarning, FailedMigrationReason, "Migration failed because target pod shutdown during migration")
 		log.Log.Object(migration).Errorf("target pod %s/%s shutdown during migration", pod.Namespace, pod.Name)
+		timeSpent = vmi.Status.MigrationState.EndTimestamp.Time.Sub(vmi.Status.MigrationState.StartTimestamp.Time)
```
@ArthurSens  `EndTimestamp` is `nil` and that's why tests are failing.

rmohr commented on pkg/monitoring/migrations/prometheus/prometheus.go:
```diff
@@ -0,0 +1,57 @@
+				Help:      "Time spent on migrations.",
+				// Will create 12 buckets, starting with migrations that take less then 2 seconds
+				// up to 2048 seconds(34,13 minutes), plus an extra for +Inf
+				Buckets: prometheus.ExponentialBuckets(2, 2, 11),
```
I find the introduction of this histogram a little bit difficult. Migrations can take a long time. I am not sure what we consider to be too long to have a dedicated bucket. @sradco did you collect such metrics in oVirt and have some insight s which would help having a good first guess for meaningful bucket sizing?

ArthurSens replied:
Yeah, I think the most appropriate approach would be to enable its configuration by the user, but that would require some API changes that I'm not familiar with yet

If you are using KubeVirt for small VMs, then you could set up small buckets... big buckets for big VMs, etc.

do you think we should introduce this metric with changes on the KubeVirt CR right from the start?

rmohr replied:
Making it configurable from the beginning may indeed be the way to go. Let's see what @sradco thinks.

sradco replied:
@rmohr Sorry, I have no experience with buckets from oVirt.

rmohr replied:
@sradco no problem, but maybe with typical migration runtimes.

rmohr replied:
Sounds to me like we should go with configurable buckets ... :/

Let's add some nice defaults, something like 1m, 5m, 10m, 30m, 1h, 2h, 3h, 4h, 5h, 6h, 7h, 10h. I don't think that more granularity makes sense for defaults. @sradco @ArthurSens what do you think?

sradco replied:
I agree

ArthurSens replied:
cool, I'll start working on it :slightly_smiling_face:

ArthurSens commented:
After a conversation offline with @danielBelenky @dhiller and @fabiand, I've re-added the target and source nodes as labels. VMI name and namespace were also added.

PR description was updated to match new changes and to talk about the API changes

ArthurSens commented:
Logic added. The virt-controller queries the KubeVirt object to get metricsConfig and then he does all the validations. It will use the default values if the configuration is missing or invalid.

With that approach, you need to restart the virt-controllers after you modify the KubeVirt object with new metricsConfig.
I think that using the virt-operator, with the informers and webhook-admitters is a better and smarter approach, but I have almost no experience with that :cry: I'll study it and refactor this part, but I'm afraid it will take some time


e2e tests are still missing

ArthurSens commented:
Oh almost forgot, after some days thinking about it, I also removed the options to create "Exponential" and "Linear" progression
I guess I added those options just because Prometheus allows us to do that. However, for the user, it's a lot easier to just declare your buckets without having to think about math. And for us, it removes a lot of validations that needed to be implemented to support those options

xpivarc commented on pkg/virt-operator/webhooks/webhook.go:
```diff
@@ -90,3 +90,82 @@ func (k *KubeVirtDeletionAdmitter) Admit(review *v1beta1.AdmissionReview) *v1bet
+	var kv *v1.KubeVirt
+	var err error
+
+	kv, err = k.client.KubeVirt(review.Request.Namespace).Get(review.Request.Name, &metav1.GetOptions{})
```
Kubevirt object should be in review.

ArthurSens replied:
Sorry, I don't think I understood what you meant here

xpivarc replied:
review -> `v1beta1.AdmissionReview`
You don't need to get Kubevirt object, just marshal it from `review.Request.Object`

xpivarc commented on pkg/virt-operator/webhooks/webhook.go:
```diff
@@ -90,3 +90,82 @@ func (k *KubeVirtDeletionAdmitter) Admit(review *v1beta1.AdmissionReview) *v1bet
+}
+
+func validateMetricsConfig(metricsConfig *v1.MetricsConfig) *v1beta1.AdmissionResponse {
+	var err error
```
No need to have it declared explicitly.

xpivarc commented on pkg/virt-operator/webhooks/webhook.go:
```diff
@@ -90,3 +90,82 @@ func (k *KubeVirtDeletionAdmitter) Admit(review *v1beta1.AdmissionReview) *v1bet
 	return validating_webhooks.NewPassingAdmissionResponse()
 }
+
+func NewKubeVirtCreationAdmitter(client kubecli.KubevirtClient) *KubeVirtCreationAdmitter {
```
You need to register this like [here](https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-operator/creation/components/webhooks.go#L47).

xpivarc commented on pkg/virt-controller/watch/application.go:
```diff
@@ -410,7 +417,24 @@ func (vca *VirtControllerApp) initCommon() {
+	// Query the KubeVirt object to get Metrics config
+	kvList, err := vca.clientSet.KubeVirt(k8sv1.NamespaceAll).List(&v1.ListOptions{})
+	if err != nil {
+		panic(err)
```
Looking at start of the method... I would use `golog.Fatal` for consistency... Maybe we can add little bit of context to this error, so the user knows what to do.

xpivarc reviewed (changes requested):
I had only a quick look. Can we also functional tests for this?

ArthurSens commented:
@xpivarc @rmohr 
The failing tests have passed locally, so I think it's ready for review

ArthurSens commented:
/retest

fabiand commented:
@sradco @yuvalturg @rmohr what should we do about this PR - drop or keep and rebase?

fabiand commented:
@dhiller do you know why this PR is not automatically closed?
